File: solution/714. Best Time to Buy and Sell Stock with Transaction Fee/714.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        
        int sold = 0;
        int hold = - prices[0];
        
        for (int price : prices) {
            int sold_old = sold;
            sold = max(hold + price - fee, sold);
            hold = max(sold_old - price, hold);
        }
        
        return sold;
    }
};

File: solution/714. Best Time to Buy and Sell Stock with Transaction Fee/714.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        hold, sold = -10**9, 0
        for x in prices:
            pre_hold = hold
            hold = max(hold, sold - x)
            sold = max(sold, pre_hold + x - fee)
        return sold
